### PR TITLE
[CALCITE-3004] RexOver is incorrectly pushed down in ProjectSetOpTran…

### PR DESCRIPTION
**Pull request by chunweilei**
…sposeRule and ProjectCorrelateTransposeRule (Chunwei Lei)

JIRA: https://issues.apache.org/jira/browse/CALCITE-3004

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectSetOpTransposeRule.java [hunk 1]
```diff
@@ -90,23 +91,35 @@ public void onMatch(RelOptRuleCall call) {
...
+        if (p.getRowType().equals(input.getRowType())) {
+          return;
+        }
+        newSetOpInputs.add(p);
+      }
+      SetOp newSetOp =
+          setOp.copy(setOp.getTraitSet(), newSetOpInputs);
+      node = pushProject.createNewProject(newSetOp, adjustments);
+    } else {
+      // push some expressions below the setop; this
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectSetOpTransposeRule.java]: There is `if..else..` in that I don't want to change the plan of some cases. For instance,  current plan of 
```
select ename,avg(empno) from 
(select * from emp as e1 union all select * from emp as e2)
```
is
```
LogicalAggregate(group=[{0}], EXPR$1=[AVG($1)])
  LogicalUnion(all=[true])
    LogicalProject(ENAME=[$1], EMPNO=[$0])
      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
    LogicalProject(ENAME=[$1], EMPNO=[$0])
      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
But if without` if..else..`, the plan becomes
```
LogicalAggregate(group=[{0}], EXPR$1=[AVG($1)])
  LogicalProject(ENAME=[$1], EMPNO=[$0])
    LogicalUnion(all=[true])
      LogicalProject(EMPNO=[$0], ENAME=[$1])
        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
      LogicalProject(EMPNO=[$0], ENAME=[$1])
        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
which seems redundant.

File: core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml [hunk 2]
```diff
@@ -9250,15 +9279,16 @@ LogicalProject(NAME=[$0], ENAME=[$2])
 ]]>
         </Resource>
     </TestCase>
-    <TestCase name="testProjectCorrelateTranspose">
+    <TestCase name="testProjectCorrelateTransposeWithOver">
         <Resource name="sql">
```
Comment (chunweilei) [inline core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: Note that `testProjectCorrelateTranspose` in `RelOptRulesTest.xml` is redundant[1][2].

[1] https://github.com/apache/calcite/blob/master/core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml#L9226
[2] https://github.com/apache/calcite/blob/master/core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml#L9253

Comment (julianhyde) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: Ah. Somehow there's a duplicate key in RelOptRulesTest.xml. Maybe it's a merge error, I don't know. Can someone add code to DiffRepository to make it fail when reading a .xml file if there are duplicate keys.

Comment (chunweilei) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: @julianhyde I would like to do it. : )

Comment (chunweilei) [reply core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml]: Opened an issue[(CALCITE-3009](https://issues.apache.org/jira/browse/CALCITE-3009)) for it.

**Comment by chunweilei** [comment]
@julianhyde @zhztheplayer, could you please review the PR? Thanks a lot.

**Comment by chunweilei** [comment]
Updated for resolving conflicts.

**Pull request merged**